go-controller/pkg/node/gateway_udn.go

```diff
@@ -0,0 +1,154 @@
+
+// UserDefinedNetworkGateway contains information
+// required to program a UDN at each node's
+// gateway.
+// NOTE: Currently invoked only for primary networks.
+type UserDefinedNetworkGateway struct {
+	// network information
+	util.NetInfo
+	// stores the networkID of this network
+	networkID int
+	// node that its programming things on
+	node *v1.Node
+}
+
+func NewUserDefinedNetworkGateway(netInfo util.NetInfo, networkID int, node *v1.Node) *UserDefinedNetworkGateway {
+	return &UserDefinedNetworkGateway{
+		NetInfo:   netInfo,
+		networkID: networkID,
+		node:      node,
+	}
+}
+
+// AddNetwork will be responsible to create all plumbings
+// required by this UDN on the gateway side
+func (udng *UserDefinedNetworkGateway) AddNetwork() error {
+	return udng.addUDNManagementPort()
+}
+
+// DelNetwork will be responsible to remove all plumbings
+// used by this UDN on the gateway side
+func (udng *UserDefinedNetworkGateway) DelNetwork() error {
+	return udng.deleteUDNManagementPort()
+}
+
+// addUDNManagementPort does the following:
+// STEP1: creates the OVS interface on br-int for the UDN's management port interface
+// STEP2: It saves the MAC address generated on the 1st go as an option on the OVS interface
+// so that it persists on reboots
+// STEP3: creates the management port netdevice on the host
+// STEP4: adds the management port IP .2 to the netdevice
+func (udng *UserDefinedNetworkGateway) addUDNManagementPort() error {
+	var err error
+	interfaceName := util.GetNetworkScopedK8sMgmtHostIntfName(uint(udng.networkID))
+	var networkLocalSubnets []*net.IPNet
+	if udng.TopologyType() == types.Layer3Topology {
+		networkLocalSubnets, err = util.ParseNodeHostSubnetAnnotation(udng.node, udng.GetNetworkName())
+		if err != nil {
+			return fmt.Errorf("waiting for node %s to start, no annotation found on node for network %s: %v",
+				udng.node.Name, udng.GetNetworkName(), err)
+		}
+	} else if udng.TopologyType() == types.Layer2Topology {
+		globalFlatL2Networks := udng.Subnets()
+		for _, globalFlatL2Network := range globalFlatL2Networks {
+			networkLocalSubnets = append(networkLocalSubnets, globalFlatL2Network.CIDR)
+		}
+	}
+	networkMTU := udng.NetInfo.MTU()
+	if networkMTU == 0 {
+		networkMTU = config.Default.MTU
+	}
+
+	// STEP1
+	stdout, stderr, err := util.RunOVSVsctl(
+		// delete if any port exists just to clean things up
+		"--", "--if-exists", "del-port", "br-int", interfaceName,
+		"--", "--may-exist", "add-port", "br-int", interfaceName,
+		"--", "set", "interface", interfaceName,
+		"type=internal", "mtu_request="+fmt.Sprintf("%d", networkMTU),
+		"external-ids:iface-id="+udng.GetNetworkScopedK8sMgmtIntfName(udng.node.Name),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to add port to br-int for network %s, stdout: %q, stderr: %q, error: %v",
+			udng.GetNetworkName(), stdout, stderr, err)
+	}
+	klog.V(3).Infof("Added OVS management port interface %s for network %s", interfaceName, udng.GetNetworkName())
+
+	// STEP2
```

Comment:
Maybe is good to fail here is `networkLocalSubnets` is empty, I think we are not checking propertly if we configure layer2 udn primary network without subnets (you can have a multi homing layer2 without subnets)

What is clear is that we should not do anything we are doing from this point if there are no local subnets.

Reply:
I'll push a PR that ensure primary UDN on L2 topologies require a subnet.

Reply:
perfect, thanks!

Reply:
here it is: #4556

Reply:
awesome so... this comment is auto-resolved thanks to Miguel's fix :)